### PR TITLE
feat: plugin parameter types service layer

### PR DESCRIPTION
**Pull request by mtrapnell-nist**
The plugin parameter type unit tests in this PR make use of the fixtures/actions/helpers that were separated out into conftest.py, actions.py and helpers.py.

### CODE REVIEW

File: tests/unit/restapi/lib/actions.py [hunk 1]
```diff
@@ -101,3 +104,32 @@ def register_queue(
         json=payload,
         follow_redirects=True,
     )
+
+
+def create_plugin_parameter_type(
+    client: FlaskClient,
+    name: str,
+    group_id: int,
```
Comment (keithmanville) [inline tests/unit/restapi/lib/actions.py:112]: should have a `description` as well. Present in ORM and added to schema by #460

File: tests/unit/restapi/v1/conftest.py [hunk 2]
```diff
@@ -90,3 +90,34 @@ def registered_queues(
         "queue2": queue2_response,
         "queue3": queue3_response,
     }
+
+
+@pytest.fixture
+def registered_plugin_parameter_types(
+    client: FlaskClient, db: SQLAlchemy, auth_account: dict[str, Any]
+) -> dict[str, Any]:
+    # TODO: Move this to conftest.py after rebasing off dev
```
Comment (keithmanville) [inline tests/unit/restapi/v1/conftest.py:99]: Looks like this happened. Comment can be removed.

File: tests/unit/restapi/v1/conftest.py [hunk 3]
```diff
@@ -90,3 +90,34 @@ def registered_queues(
...
+    plugin_param_type1_response = actions.create_plugin_parameter_type(
+        client,
+        name="int",
+        group_id=auth_account["default_group_id"],
+        structure="",
+    )
```
Comment (keithmanville) [inline tests/unit/restapi/v1/conftest.py:105]: actions return `TestResponse` but the fixture should return a `dict`, so append `.get_json()` to each `create_plugin_parameter_type` call here.

File: tests/unit/restapi/v1/test_pluginparametertypes.py [hunk 4]
```diff
@@ -0,0 +1,817 @@
...
+
+from dioptra.restapi.routes import V1_PLUGIN_PARAMETER_TYPES_ROUTE, V1_ROOT
+
+from ..lib import actions, helpers
+
+# -- Actions ---------------------------------------------------------------------------
+
+
+def modify_plugin_parameter_type(
+    client: FlaskClient, id: int, new_name: str, new_structure: dict[str, Any]
```
Comment (keithmanville) [inline tests/unit/restapi/v1/test_pluginparametertypes.py:39]: add `description`

File: tests/unit/restapi/v1/test_pluginparametertypes.py [hunk 5]
```diff
@@ -0,0 +1,817 @@
...
+
+    This test validates this scenario by following these actions:
+
+    - The user is able to retrieve a list of all plugin parameter types.
+    - The returned list of plugin parameter types match the full list of existing plugin
+      parameter types.
+    """
+    plugin_param_type1_expected = registered_plugin_parameter_types[
+        "plugin_param_type1"
+    ].get_json()
```
Comment (keithmanville) [inline tests/unit/restapi/v1/test_pluginparametertypes.py:421]: `.get_json()` can be removed once fixture is update.

Comment (keithmanville) [reply tests/unit/restapi/v1/test_pluginparametertypes.py:421]: Needs to be updated throughout the tests.

File: tests/unit/restapi/v1/test_pluginparametertypes.py [hunk 6]
```diff
@@ -0,0 +1,817 @@
...
+    plugin_param_type2 = registered_plugin_parameter_types[
+        "plugin_param_type2"
+    ].get_json()
+    plugin_param_type3 = registered_plugin_parameter_types[
+        "plugin_param_type3"
+    ].get_json()
+
+    start_name = plugin_param_type1["id"]
```
Comment (keithmanville) [inline tests/unit/restapi/v1/test_pluginparametertypes.py:737]: should be "name" not "id"

File: tests/unit/restapi/v1/test_pluginparametertypes.py [hunk 7]
```diff
@@ -0,0 +1,817 @@
...
+    - This causes a 400 error to be returned.
+    """
+    existing_plugin_param_type = registered_plugin_parameter_types[
+        "plugin_param_type1"
+    ].get_json()
+
+    assert_cannot_create_existing_plugin_parameter_type(
+        client,
+        name=existing_plugin_param_type,
+        group_id=auth_account["default_group_id"],
```
Comment (keithmanville) [inline tests/unit/restapi/v1/test_pluginparametertypes.py:803]: need to access this through `registered_users`

File: tests/unit/restapi/v1/test_pluginparametertypes.py [hunk 8]
```diff
@@ -0,0 +1,817 @@
...
+    plugin_param_type_expected_list = [
+        plugin_param_type1_expected,
+        plugin_param_type2_expected,
+        plugin_param_type3_expected,
+    ]
+
+    assert_retrieving_plugin_parameter_types_works(
+        client,
+        expected=plugin_param_type_expected_list,
+        group_id=auth_account["default_group_id"],
```
Comment (keithmanville) [inline tests/unit/restapi/v1/test_pluginparametertypes.py:520]: use first `registered_user`

File: tests/unit/restapi/v1/test_pluginparametertypes.py (hunk 7 above)
Comment (keithmanville) [reply tests/unit/restapi/v1/test_pluginparametertypes.py:803]: @jkglasbrenner We changed `auth_account` to return the response from login. Do we want to change it back? Makes more sense for many of these tests that need the currently logged in user to not need to go through the full list of registered users.

**Review by keithmanville — CHANGES_REQUESTED** [review]
I think this is in a state where we can start to develop the service layer against these tests.

File: tests/unit/restapi/lib/actions.py [hunk 9]
```diff
@@ -101,3 +104,32 @@ def register_queue(
         json=payload,
         follow_redirects=True,
     )
+
+
+def create_plugin_parameter_type(
```
Comment (andrewhand) [inline tests/unit/restapi/lib/actions.py:109]:
```suggestion
def register_plugin_parameter_type(
```
Change create to register to keep established naming schema.

File: tests/unit/restapi/lib/actions.py [hunk 10]
```diff
@@ -101,3 +104,32 @@ def register_queue(
...
+    client: FlaskClient,
+    name: str,
+    group_id: int,
+    structure: dict[str, Any] | None = None,
```
Comment (andrewhand) [inline tests/unit/restapi/lib/actions.py:113]: > should have a `description` as well. Present in ORM and added to schema by #460

```suggestion
    client: FlaskClient,
    name: str,
    group_id: int,
    structure: dict[str, Any] | None = None,
    description: str | None = None,
```
Adding description as mentioned.

File: tests/unit/restapi/lib/actions.py [hunk 11]
```diff
@@ -101,3 +104,32 @@ def register_queue(
...
+        group_id: The group to create the new plugin parameter type in.
+        structure: Optional JSON-type field for further constraining a type's structure.
+
+    Returns:
+        The response from the API.
+    """
+    payload = {"name": name, "group_id": group_id}
+
+    if structure:
+        payload["structure"] = structure
```
Comment (andrewhand) [inline tests/unit/restapi/lib/actions.py:129]: > should have a `description` as well. Present in ORM and added to schema by #460
```suggestion
    payload = {"name": name, "group_id": group_id}

    if structure:
        payload["structure"] = structure
        
    if description:
        payload["description"] = description
```
Adding description to payload

File: tests/unit/restapi/v1/conftest.py [hunk 12]
```diff
@@ -90,3 +90,34 @@ def registered_queues(
...
         "queue3": queue3_response,
     }
+
+
+@pytest.fixture
+def registered_plugin_parameter_types(
+    client: FlaskClient, db: SQLAlchemy, auth_account: dict[str, Any]
+) -> dict[str, Any]:
+    # TODO: Move this to conftest.py after rebasing off dev
+    plugin_param_type1_response = actions.create_plugin_parameter_type(
```
Comment (andrewhand) [inline tests/unit/restapi/v1/conftest.py:100]:
```suggestion
    plugin_param_type1_response = actions.register_plugin_parameter_type(
```
Change create to register to keep established naming schema.

File: tests/unit/restapi/v1/conftest.py [hunk 13]
```diff
@@ -90,3 +90,34 @@ def registered_queues(
...
+        client,
+        name="int",
+        group_id=auth_account["default_group_id"],
+        structure="",
```
Comment (andrewhand) [inline tests/unit/restapi/v1/conftest.py:104]:
```suggestion
        client,
        name="int",
        group_id=auth_account["default_group_id"],
        structure="",
        description="The first parameter type.",
```
Adding description that is queried in tests.

File: tests/unit/restapi/v1/conftest.py (hunk 3 above)
Comment (andrewhand) [inline tests/unit/restapi/v1/conftest.py:105]: > actions return `TestResponse` but the fixture should return a `dict`, so append `.get_json()` to each `create_plugin_parameter_type` call here.
```suggestion
    ).get_json()
```

File: tests/unit/restapi/v1/conftest.py [hunk 14]
```diff
@@ -90,3 +90,34 @@ def registered_queues(
...
+    )
+    plugin_param_type2_response = actions.create_plugin_parameter_type(
+        client,
+        name="integer",
+        group_id=auth_account["default_group_id"],
+        structure="",
+    )
```
Comment (andrewhand) [inline tests/unit/restapi/v1/conftest.py:111]: > actions return `TestResponse` but the fixture should return a `dict`, so append `.get_json()` to each `create_plugin_parameter_type` call here.
```suggestion
    ).get_json()
```

File: tests/unit/restapi/v1/conftest.py [hunk 15]
```diff
@@ -90,3 +90,34 @@ def registered_queues(
...
+    # This is intentionally named using a different pattern for search query testing
+    plugin_param_type3_response = actions.create_plugin_parameter_type(
+        client,
+        name="string",
+        group_id=auth_account["default_group_id"],
+        structure="",
+    )
```
Comment (andrewhand) [inline tests/unit/restapi/v1/conftest.py:118]: > actions return `TestResponse` but the fixture should return a `dict`, so append `.get_json()` to each `create_plugin_parameter_type` call here.
```suggestion
    ).get_json()
```

File: tests/unit/restapi/v1/conftest.py [hunk 16]
```diff
@@ -90,3 +90,34 @@ def registered_queues(
...
+        name="integer",
+        group_id=auth_account["default_group_id"],
+        structure="",
+    )
+    # This is intentionally named using a different pattern for search query testing
+    plugin_param_type3_response = actions.create_plugin_parameter_type(
+        client,
+        name="string",
+        group_id=auth_account["default_group_id"],
+        structure="",
```
Comment (andrewhand) [inline tests/unit/restapi/v1/conftest.py:117]:
```suggestion
        client,
        name="string",
        group_id=auth_account["default_group_id"],
        structure="",
        description="Not retrieved.",
```
Add description for a param type that isn't queried.

File: tests/unit/restapi/v1/conftest.py [hunk 17]
```diff
@@ -90,3 +90,34 @@ def registered_queues(
...
+        client,
+        name="int",
+        group_id=auth_account["default_group_id"],
+        structure="",
+    )
+    plugin_param_type2_response = actions.create_plugin_parameter_type(
+        client,
+        name="integer",
+        group_id=auth_account["default_group_id"],
+        structure="",
```
Comment (andrewhand) [inline tests/unit/restapi/v1/conftest.py:110]:
```suggestion
        client,
        name="integer",
        group_id=auth_account["default_group_id"],
        structure="",
        description="The second parameter type.",
```
Adding description that is queried in tests.

**Review by andrewhand — CHANGES_REQUESTED** [review]
Just a review of actions and conftest needed to merge these files into dev for use in Plugin tests. Did not review the test files for this endpoint.

**Comment by mtrapnell-nist** [comment]
@andrewhand  I applied all the necessary changes for conftest.py and actions.py, so they should be ready for you to use on your branch.

**Comment by mtrapnell-nist** [comment]
@keithmanville 
This includes all changes for the service layer for plugin parameter types and is ready for review.

**Pull request merged**